### PR TITLE
Refunds

### PR DESCRIPTION
**Pull request by KENDARRR**
## Brief Description

Started with the refund api, but ran into too many issues with the checkout api to test refunds without some alterations. I fixed the checkout page to include a stripe session redirect. Added a webhook so upon successful payment the server is notified and payment intent is saved.

Issue #77

- [ !] Bug fix
- [ !] Feature
- [ !] Major change (potentially breaking, list anything that is broken below)
- [ ] Documentation

## Testing done

Stripe CLI webhook testing was done. A lot of this had to be done through the stripe dashboard.

- [ !] Docker was used.
- [ ] Docker was not used and I have listed my configuration below.

**System Specifications**:
* Operating System: Win 10
* Node Version: 16.16.0
* NPM Version: 8.18.0

## Checklist:

- [ !] My code follows the styling guidelines.
- [ ] I have added unit tests and verified that they pass.
- [ !] I have used the linter and fixed any linting issues.
- [ !] I have commented the code.
- [ !] I have adjusted the documentation to match the changed code.
- [ !] Prior to submitting the PR, I made sure the latest changes were merged with my branch.
- [ !] Existing tests pass locally with changes.

### Explanation for unchecked

Testing was done through stripe CLI so unit tests weren't implemented yet

### CODE REVIEW

File: client/src/components/Ticketing/checkout/CheckoutPage.tsx [hunk 1]
```diff
@@ -19,9 +19,10 @@ import CompleteOrderForm, {CheckoutFormInfo} from './CompleteOrderForm';
 import {selectDonation} from '../ticketingmanager/donationSlice';
 import {useNavigate} from 'react-router-dom';
 
-// FYI this is ok to stay here; it's the public key so other people can't do anything with it anyway.
-// Replace this with your stripe public key
-const stripePromise = loadStripe(process.env.REACT_APP_PUBLIC_STRIPE_KEY);
+//const pk = `${process.env.PUBLIC_STRIPE_KEY}`;
+const pk = `pk_test_51LarKnLnmVGBW71PEQcPH5MivNcFakrZR5hg3jRtSPIp58ziwNHjFfOVktgxWdMD8OHdteqMWiYh1MWCTOY5QcLH00BS27YNOM`;
+//const pk = `pk_live_51LarKnLnmVGBW71PE5iZvRz1YyozIa1DggXkHwUB6xoVwaEhohEZKsestQwUtx0jAl98f3As5T7wIJ7MGDsvHPik000tLSC9xC`;
+const stripePromise = loadStripe(pk);
```
Comment (ben16williams) [inline client/src/components/Ticketing/checkout/CheckoutPage.tsx:25]: I know it's just a public key, but still would be nice to access as an environment variable. Not a blocker

File: server/src/api/events/event.router.ts [hunk 2]
```diff
@@ -12,16 +11,18 @@ import {checkIn,
...
+} from './event.service';
 import {checkJwt, checkScopes} from '../../auth';
 export const eventRouter = Router();
 
-const stripeKey = process.env.PRIVATE_STRIPE_KEY ?
-  process.env.PRIVATE_STRIPE_KEY : '';
+const stripeKey = `${process.env.PRIVATE_STRIPE_KEY}`;
+const stripe = require('stripe')(stripeKey);
+//const webhookKey = `${process.env.PRIVATE_STRIPE_WEBHOOK}`;
+const webhookKey = `whsec_f9a21f44f3657e3f840903f4260adbd74f74ea9940e405ac19466e58ecb60b82`;
```
Comment (ben16williams) [inline server/src/api/events/event.router.ts:23]: This should probably be accessed by an environment variable right?

File: server/src/api/events/event.router.ts [hunk 3]
```diff
@@ -206,9 +209,30 @@ eventRouter.post('/checkout', async (req: Request, res: Response) => {
...
+      } catch (err: any) {
+        console.error(err.message);
+        throw new Error('Cost Calculation Error');
+      };
+    };
+    console.log(data);
     const session = await stripe.checkout.sessions.create({
       payment_method_types: ['card'],
-      // this is the offending area
       // all this stuff needs to be replaced by info from DB based on event ID
```
Comment (ben16williams) [inline server/src/api/events/event.router.ts:236]: Can this be removed as well? Might be confusing later on if it's left in

File: server/src/api/events/event.router.ts [hunk 4]
```diff
@@ -232,19 +255,64 @@ eventRouter.post('/checkout', async (req: Request, res: Response) => {
...
+  }
+
+  console.log('webhook passed');
+  console.log(event);
+
+  try{
+    console.log('starting fulfillment');
+    const session = event.data.object;
+    console.log('session = ' + session);
+    const amount = session.amount/100];
```
Comment (ben16williams) [inline server/src/api/events/event.router.ts:302]: extra bracket

**Review by ben16williams — DISMISSED** [review]

**Review by ben16williams — APPROVED** [review]

**Comment by KENDARRR** [comment]
Refunds, check out, and webhook added.

**Pull request merged**